File: raspberry-pi/backend/app/schemas/module.py
"""Module schemas."""

from datetime import datetime
from pydantic import BaseModel, model_validator


class ModuleConnectivityResponse(BaseModel):
    """Module connectivity response schema."""

    isOnline: bool
    lastSeen: datetime | None = None

    @model_validator(mode='after')
    def check_online_requires_last_seen(self):
        if self.isOnline and self.lastSeen is None:
            raise ValueError("If module is online, lastSeen cannot be None")
        return self


# Module Responses
class ModuleResponse(BaseModel):
    """Module response schema."""

    id: str
    coupled: bool
    coupledPlantId: int | None = None
    connectivity: ModuleConnectivityResponse

File: central-hub/iot-platform/frontend/app/lib/types/module.ts
// Module connectivity type definition
export type ModuleConnectivity =
  | {
      isOnline: true
      lastSeen: string
    }
  | {
      isOnline: false
      lastSeen: string | null
    }

// Module type definition
export type Module = {
  id: string
  connectivity: ModuleConnectivity
  coupled: boolean
  coupledPlantId: number | null
}
